proto/rdp/README.md

```diff
@@ -0,0 +1,277 @@
+
+Problems:
+- No support for diagnostics for code range. It only supports start position.
+- No support for code suggestions (also known as auto-correct or fix).
+- It's ..... XML. It's true that some diagnostic tools support checkstyle
+format, but not everyone wants to support it.
+- The checkstyle itself is actually a diagnostic tool for Java and its
+  output format is actually not well-documented and not meant to be
+  used as generic format. Some linters just happens to use the same format(?).
+
+## Background: Altenatives
+
+There are altenative solutions out there (which are not used by reviewdog) as
+well.
+
+### *Problem Matcher*
+[VSCode](https://vscode-docs.readthedocs.io/en/stable/editor/tasks/#defining-a-problem-matcher)
+and [GitHub Actions](https://github.com/actions/toolkit/blob/master/docs/problem-matchers.md)
+uses [Problem Matcher](https://github.com/actions/toolkit/blob/master/docs/problem-matchers.md)
+to support arbitrary diagnostic tools. It's similar to errorformat, but it uses regex.
+
+Problems:
+- No support for code suggestions (also known as auto-correct or fix).
+- Output format of matched results are undocumented and it seems to be used internally in VSCode and GitHub Actions.
+- It's hard to write problem matchers for complicated output.
+
+### *Language Server Protocol (LSP)*
+[Language Server Protocol Specification](https://microsoft.github.io/language-server-protocol/specifications/specification-current/)
+
+LSP supports [Diagnostic](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#diagnostic)
+to represents a diagnostic, such as a compiler error or warning.
+It's great for editor integration and is widely used these days as well.
+RDP message is actually inspired by LSP Diagnostic message too.
+
+Problems:
+- LSP and the Diagnostic message is basically per one file. It's not always
+  suited to be used as diagnostic tools output because they often need to
+  report diagnostic results for multiple files and outputing json per file does
+  not make very much sense.
+- LSP's Diagnostic message doesn't have code suggestions (code action) data.
+  Code action have data about associated diagnostic on the contrary and the
+  code action message itself doesn't contain text edit data too, so LSP's
+  messages are not suited to represent a diagnosis result with suggested fix.
+- Unnatural position representation: Position in LSP are zero-based and
+  character offset is based on [UTF-16 code units](https://github.com/microsoft/language-server-protocol/issues/376).
+  These are not widely used by diagnostic tools, development tools nor code
+  review API such as GitHub, GitLab and Gerrit....
+  In addition, UTF-8 is defact-standard of text file encoding as well these days.
+
+## Reviewdog Diagnostic Protocol Concept
+Again, the idea behind the Reviewdog Diagnostic Protocol (RDP) is to standardize
+the protocol for how diagnostic tools (e.g. compilers, linters, etc..) and
```

Comment:
In addition to the language tool complaint, as a general rule you should not do `e.g. ..., etc.` -- `e.g.` is used to introduce examples, but `etc.` (*[et cetera](https://dictionary.cambridge.org/dictionary/english/etc)*) is for *and others*, which aren't really *an example*.

Here, the sample solution is to drop ` etc..` entirely. Possibly adding `and` or `or` before `code review API`.
Note: there's no need for the trailing `.` in this, as the parenthetical isn't a sentence.